Introduce DICE with Console

nupplaphil opened the pull request:
This a Draft PR.

Details: https://friendica.philipp.info/display/f1417ed3-525c-eedc-3119-682437319334
I'd like to show you an implementation of the most minimal Dependency Injection(DI)-library, I found so far.


I choose this library: https://github.com/Level-2/Dice
- It has no dependencies
- It is usable with **one** 250-lines of code PHP-File (`library/Dice.php`)
- Performance [measures](https://rawgit.com/TomBZombie/php-dependency-injection-benchmarks/master/test6_results.html)

The goal for Friendica:
- Get rid of `DependencyFactory`
- Automatically load `App` in case we need it (prerequisite for https://github.com/friendica/friendica/issues/6998#issuecomment-482520207 , because here we shouldn't load it!)
- Construct the "dependency tree" out of the box
- Prerequisite for further test-optimized refactorings
- **Remove the whole `App::__construct()` loading hell**

The PR is already tested with with console/index/worker/daemon and would run fine.

Some current downsides:
- Moved `DBAFactory::init()` into `ConfigFactory::createConfig()` because it's currently impossible for a DI-library to load it automatically
- Moved `ConfigFactory::CreatePConfig()` into `App::__construct()` because it's currently impossible for a DI-library to load it automatically
- Moved `LoggerFactory::CreateDev()` into `App::__construct()` because it's currently impossible for a DI-library to load it automatically

nupplaphil committed:
Introduce DICE with Console

nupplaphil commented:
This would be the new "static" dependency config: https://github.com/friendica/friendica/pull/7087/files#diff-83b65e470193ab648b039dcb1f2e4896

Every call, which isn't "obvious" (like `new Class()`) should be configurated here.
f.e. `new App\Router()` doesn't need to be defined, because there is no complex dependency.

MrPetovan commented on src/Core/Console.php:
```diff
@@ -15,6 +16,14 @@ class Console extends \Asika\SimpleConsole\Console
 
+	private $dice;
+
+	public function __construct(Dice $dice, $argv = null)
```
I thought it was an anti-pattern to pass the Dependency Injection Container as a class parameter?

nupplaphil replied:
You're right, but I've to start at a place.
It seems for me that there are three central "controllers", which are used to determine the initial class we need:
- `Friendica\Core\Console` => Determine the `Friendica\Console\*` class
- `Friendica\Core\Worker` => Determine the `Friendica\Worker\*` class
- `Friendica\App\Router` => Determine the `Friendica\Module\*` class

Every other classes "could" be loaded through the determined class in a later stage.

MrPetovan replied:
So, reading further I found out that it's okay to pass the DI container to a class that will act as a Composition Root like I believe the `Core\Console` class is.

MrPetovan commented on static/dependencies.conf.php:
```diff
@@ -0,0 +1,61 @@
+use Friendica\Util;
+use Psr\Log\LoggerInterface;
+
+return [
```
We'll need some serious comments here to explain each individual rule. Not only it can save someone the need to read [the entire Dice documentation](https://r.je/dice), but it'll also help adding new rules for similar use cases.

nupplaphil replied:
Yep, you're right!
As far as I can see, there shouldn't be THAT complex configuration after the initial core classes needed. Most of the others would be loaded without additional config.

MrPetovan commented on library/Dice.php:
```diff
@@ -0,0 +1,242 @@
+<?php
```
Please don't add an external library file in `library/`. This is a legacy artifact because I couldn't find any similar Composer libraries for ASN1 and OAuth1.

But this is a proper Composer dependency and should be added through this channel as well.

nupplaphil replied:
There is no composer library available for it.
https://github.com/Level-2/Dice#installation
The author vaunts that there is no "complex" dependency necessary, just including the file.

MrPetovan replied:
https://packagist.org/packages/level-2/dice

To bump the Friendica hard minimum PHP version to 7, run `composer require level-2/dice`.

Otherwise, you'll need to run `composer require level-2/dice ^2|^3` on a PHP 5 install.

nupplaphil replied:
... I'm sorry, I was too stupid to find it ^^

MrPetovan replied:
We don't need to keep the PHP 5 compatibility anymore. `develop` can require PHP 7 from now on and the 2019.06 version can require PHP 7.

nupplaphil commented:
Based on https://github.com/friendica/friendica/pull/7087/commits/449761a7e2188df0be187d9d771fbe063b457cbc , an advantage I want to show:

- We now doesn't load `App` anymore for this specific task
- DICE checks that the constructor of `Config` needs `Configuration` and `Mode` and builds it. No further logic is necessary

nupplaphil commented:
Now requires #7110 with the recent push, but it's now working with composer :-)
=> I had to choose "dev-master" because a recent bugfix is necessary to work with our factories

MrPetovan reviewed (changes requested):
You probably need to bump the minimum PHP version in `composer.json` now.

nupplaphil commented:
So, that's it for now.. The next step would be to look for the PHP 7.0 compose files ;-)

As said, I don't think that we'll need that more "complex" dependency configurations. Mainly they are used in case of "decissions" (= Factory).

nupplaphil commented:
> 
> 
> You probably need to bump the minimum PHP version in `composer.json` now.

First, I'll have to look for the compose files, after it I'd rebase this PR

nupplaphil committed:
Moving from DBA to Database
- Fixing tests & daemon
- Add new library paragonie/hidden-string
- Bump paragonie/constant_time_encoding from 1.0.4 to 2.2.3
- Hide values of password keys through configcache and environment variables
- update composer.lock

nupplaphil commented:
Edit .. rebased on 2019.06-rc including Pager/Database changes from #7231 and #7232 
=> Details tomorrow ^^

MrPetovan commented:
What a tease!

nupplaphil committed:
Refactor using DI-library
- Remove separate processes for App\Mode testing

nupplaphil committed:
Refactor SystemTest without DB

MrPetovan commented:
This PR needs to be reopened against `develop`.
